skiko/src/jvmMain/cpp/common/shaper/Shaper.cc

```diff
@@ -88,32 +93,37 @@ extern "C" JNIEXPORT jlong JNICALL Java_org_jetbrains_skia_shaper_ShaperKt__1nSh
     SkTextBlobBuilderRunHandler rh(text.c_str(), {offsetX, offsetY});
     instance->shape(text.c_str(), text.size(), *fontRunIter, *bidiRunIter, *scriptRunIter, *languageRunIter, features.data(), features.size(), width, &rh);
     SkTextBlob* blob = rh.makeBlob().release();
-    
+
     return reinterpret_cast<jlong>(blob);
 }
 
 extern "C" JNIEXPORT jlong JNICALL Java_org_jetbrains_skia_shaper_ShaperKt__1nShapeLine
-  (JNIEnv* env, jclass jclass, jlong ptr, jstring textObj, jlong fontPtr, jobject opts) {
+  (JNIEnv* env, jclass jclass, jlong ptr, jlong textPtr, jlong fontPtr, jint optsFeaturesLen, jintArray optsFeatures, jint optsBooleanProps) {
     SkShaper* instance = reinterpret_cast<SkShaper*>(static_cast<uintptr_t>(ptr));
 
-    SkString text = skString(env, textObj);
+    SkString text = *(reinterpret_cast<SkString*>(static_cast<uintptr_t>(textPtr)));
```

Comment:
Yep, `SkString& text` would read better in cases like that.

Reply:
fixed